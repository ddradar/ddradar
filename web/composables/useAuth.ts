import type { CurrentUserInfo } from '~~/server/api/v1/user/index.get'
import { type ClientPrincipal, useClientPrincipal } from '~~/server/utils/auth'

export default async function useAuth() {
  const auth = useState<ClientPrincipal | null>('auth')
  const user = useState<CurrentUserInfo | null>('user')

  auth.value = process.server
    ? useClientPrincipal(useRequestHeaders())
    : (await $fetch<{ clientPrincipal: ClientPrincipal | null }>('/.auth/me'))
        .clientPrincipal
  try {
    user.value = auth.value
      ? await $fetch('/api/v1/user', {
          headers: useRequestHeaders(['x-ms-client-principal']),
        })
      : null
  } catch (error) {
    user.value = null
  }

  const computedAuth = computed(() => auth.value)
  const computedUser = computed(() => user.value)
  const id = computed(() => user.value?.id)
  const name = computed(() => user.value?.name)
  const isLoggedIn = computed(() => !!user.value)
  const isAdmin = computed(
    () => !!auth.value?.userRoles.includes('administrator')
  )

  const saveUser = async (body: CurrentUserInfo) => {
    user.value = await $fetch('/api/v1/user', { method: 'POST', body })
  }

  return {
    auth: computedAuth,
    user: computedUser,
    id,
    name,
    isLoggedIn,
    isAdmin,
    saveUser,
  }
}
