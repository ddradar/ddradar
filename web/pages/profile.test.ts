import { publicUser } from '@ddradar/core/test/data'
import Oruga, { useProgrammatic } from '@oruga-ui/oruga-next'
import { bulmaConfig } from '@oruga-ui/theme-bulma'
import { beforeEach, describe, expect, test, vi } from 'vitest'
import { createI18n } from 'vue-i18n'

import { CurrentUserInfo } from '~/server/api/v1/user/index.get'
import useAuth from '~~/composables/useAuth'
import Page from '~~/pages/profile.vue'
import { mountAsync } from '~~/test/test-utils'

const open = vi.fn()
vi.mock('~~/composables/useAuth')
vi.mock('@oruga-ui/oruga-next', async () => {
  const actual = (await vi.importActual(
    '@oruga-ui/oruga-next'
  )) as typeof import('@oruga-ui/oruga-next')
  return { ...actual, useProgrammatic: vi.fn() }
})
vi.mocked(useProgrammatic).mockReturnValue({
  oruga: { notification: { open } },
})

describe('Page /profile', () => {
  const i18n = createI18n({ legacy: false, locale: 'ja' })

  describe.each(['ja', 'en'])('{ locale: "%s" } snapshot test', locale => {
    const i18n = createI18n({ legacy: false, locale })

    test('renders correctly', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked(useAuth).mockResolvedValue({ user: ref(null) } as any)
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      const wrapper = await mountAsync(Page, {
        global: { plugins: [[Oruga, bulmaConfig], i18n] },
      })

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('isNewUser()', () => {
    test.each([
      [null, true],
      [publicUser, false],
    ])('{ user: %o } returns %o', async (user, isNewUser) => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked(useAuth).mockResolvedValue({ user: ref(user) } as any)
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      const wrapper = await mountAsync(Page, {
        global: { plugins: [[Oruga, bulmaConfig], i18n] },
      })

      // Assert
      const vm = wrapper.getComponent(Page).vm as unknown as {
        isNewUser: boolean
      }
      expect(vm.isNewUser).toBe(isNewUser)
    })
  })
  describe('hasError()', async () => {
    const user = {
      id: publicUser.id,
      name: publicUser.name,
      area: publicUser.area,
      code: publicUser.code,
      variant: '',
    }
    /* eslint-disable @typescript-eslint/no-explicit-any */
    vi.mocked(useAuth).mockResolvedValue({ user: ref(null) } as any)
    /* eslint-enable @typescript-eslint/no-explicit-any */
    const wrapper = await mountAsync(Page, {
      global: { plugins: [[Oruga, bulmaConfig], i18n] },
    })
    test.each([
      { ...user, variant: 'danger' },
      { ...user, id: '@' },
      { ...user, name: null },
      { ...user, area: -1 },
      { ...user, code: 1.3 },
      { ...user, code: 100000000 },
    ])('%o returns true', async diff => {
      // Arrange - Act
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const vm = wrapper.getComponent(Page).vm as any
      Object.entries(diff).map(([key, value]) => (vm[key] = value))

      // Assert
      expect(vm.hasError).toBe(true)
    })
    test.each([
      { ...user },
      { ...user, id: 'FOO' },
      { ...user, variant: 'success' },
      { ...user, code: null },
    ])('%o returns false', async user => {
      // Arrange - Act
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const vm = wrapper.getComponent(Page).vm as any
      Object.entries(user).map(([key, value]) => (vm[key] = value))

      // Assert
      expect(vm.hasError).toBe(false)
    })
  })

  describe('checkId()', async () => {
    /* eslint-disable @typescript-eslint/no-explicit-any */
    vi.mocked(useAuth).mockResolvedValue({ user: ref(null) } as any)
    /* eslint-enable @typescript-eslint/no-explicit-any */
    const wrapper = await mountAsync(Page, {
      global: { plugins: [[Oruga, bulmaConfig], i18n] },
    })
    const vm = wrapper.getComponent(Page).vm as unknown as {
      checkId(): Promise<void>
      id: string
      variant: string
      message: string
    }
    test.each(['', '@'])('{ id: "%s" } sets error', async id => {
      // Arrange- Act
      vm.id = id
      await vm.checkId()

      // Assert
      expect(vm.variant).toBe('danger')
      expect(vm.message).not.toBe('')
    })
    test('{ id: <Duplicated ID> } sets error', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked($fetch).mockResolvedValue({ exists: true })
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      vm.id = publicUser.id
      await vm.checkId()

      // Assert
      expect(vm.variant).toBe('danger')
      expect(vm.message).not.toBe('')
    })
    test('{ id: <Avaliable ID> } does not set error', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked($fetch).mockResolvedValue({ exists: false })
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      vm.id = publicUser.id
      await vm.checkId()

      // Assert
      expect(vm.variant).toBe('success')
      expect(vm.message).not.toBe('')
    })
    test('sets type: "" if API call failed', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked($fetch).mockRejectedValue('Error')
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      vm.id = publicUser.id
      await vm.checkId()

      // Assert
      expect(vm.variant).toBe('')
    })
  })
  describe('save()', async () => {
    const saveUser = vi.fn<[CurrentUserInfo], Promise<void>>()
    /* eslint-disable @typescript-eslint/no-explicit-any */
    vi.mocked(useAuth).mockResolvedValue({
      user: ref(publicUser),
      saveUser,
    } as any)
    /* eslint-enable @typescript-eslint/no-explicit-any */
    const wrapper = await mountAsync(Page, {
      global: { plugins: [[Oruga, bulmaConfig], i18n] },
    })
    const vm = wrapper.getComponent(Page).vm as unknown as {
      save(): Promise<void>
      id: string
      variant: string
      message: string
    }

    beforeEach(() => open.mockClear())

    test('calls saveUser() and notifies Success', async () => {
      // Arrange - Act
      await vm.save()

      // Assert
      expect(saveUser).toBeCalled()
      expect(open).toBeCalledWith({
        message: '保存しました',
        variant: 'success',
        position: 'top',
      })
    })
    test('notifies Error if saveUser() throws error', async () => {
      // Arrange
      saveUser.mockRejectedValue('Error')
      // Act
      await vm.save()

      // Assert
      expect(saveUser).toBeCalled()
      expect(open).toBeCalledWith({
        message: 'Error',
        variant: 'danger',
        position: 'top',
      })
    })
  })
})
