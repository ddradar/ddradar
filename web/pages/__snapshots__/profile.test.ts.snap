// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Page /profile > { locale: "en" } snapshot test > renders correctly 1`] = `
<section
  class="section"
>
  <transition-stub
    appear="false"
    css="true"
    name="fade"
    persisted="false"
  >
    <article
      class="notification is-danger"
    >
      <!--v-if-->
      <!--v-if-->
      <div
        class="media"
      >
        <span
          aria-hidden=""
          class="icon is-large media-left"
        >
          <i
            class="mdi mdi-alert-circle mdi-48px"
          />
        </span>
        <div
          class="media-content"
        >
          
          User registration has not been completed yet. Please continue to enter the following information.
          
        </div>
      </div>
    </article>
  </transition-stub>
  <h1
    class="title"
  >
    User Settings
  </h1>
  <div
    class="field"
  >
    
    <label
      class="label"
    >
      
      ID 
      <div
        class="b-tooltip is-dark is-right"
      >
        <transition-stub
          appear="false"
          css="true"
          name="fade"
          persisted="false"
        >
          <div
            class="tooltip-content is-right is-dark"
            style="display: none;"
          >
            <span
              class=""
            />
            
            can not be changed after registration.
            
          </div>
        </transition-stub>
        <div
          class="tooltip-trigger"
        >
          
          <span
            class="icon is-small"
          >
            <i
              class="mdi mdi-help-circle-outline"
            />
          </span>
          
        </div>
      </div>
      
    </label>
    
    
    <div
      class="control"
    >
      <input
        autocomplete="off"
        class="input"
        loading="false"
        placeholder="ID"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="field"
  >
    
    <label
      class="label"
    >
      
      Display Name
      
    </label>
    
    
    <div
      class="control"
    >
      <input
        autocomplete="off"
        class="input"
        required=""
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="field"
  >
    
    <label
      class="label"
    >
      
      Area 
      <div
        class="b-tooltip is-dark is-right"
      >
        <transition-stub
          appear="false"
          css="true"
          name="fade"
          persisted="false"
        >
          <div
            class="tooltip-content is-right is-dark"
            style="display: none;"
          >
            <span
              class=""
            />
            
            can not be changed after registration.
            
          </div>
        </transition-stub>
        <div
          class="tooltip-trigger"
        >
          
          <span
            class="icon is-small"
          >
            <i
              class="mdi mdi-help-circle-outline"
            />
          </span>
          
        </div>
      </div>
      
    </label>
    
    
    <div
      class="control select"
    >
      <select
        class=""
        modelmodifiers="[object Object]"
      >
        
        <!--v-if-->
        
        
        
        <option
          value="0"
        >
          undefined
        </option>
        <option
          value="1"
        >
          Hokkaido
        </option>
        <option
          value="2"
        >
          Aomori
        </option>
        <option
          value="3"
        >
          Iwate
        </option>
        <option
          value="4"
        >
          Miyagi
        </option>
        <option
          value="5"
        >
          Akita
        </option>
        <option
          value="6"
        >
          Yamagata
        </option>
        <option
          value="7"
        >
          Fukushima
        </option>
        <option
          value="8"
        >
          Ibaraki
        </option>
        <option
          value="9"
        >
          Tochigi
        </option>
        <option
          value="10"
        >
          Gunma
        </option>
        <option
          value="11"
        >
          Saitama
        </option>
        <option
          value="12"
        >
          Chiba
        </option>
        <option
          value="13"
        >
          Tokyo
        </option>
        <option
          value="14"
        >
          Kanagawa
        </option>
        <option
          value="15"
        >
          Niigata
        </option>
        <option
          value="16"
        >
          Toyama
        </option>
        <option
          value="17"
        >
          Ishikawa
        </option>
        <option
          value="18"
        >
          Gukui
        </option>
        <option
          value="19"
        >
          Yamanashi
        </option>
        <option
          value="20"
        >
          Nagano
        </option>
        <option
          value="21"
        >
          Gifu
        </option>
        <option
          value="22"
        >
          Shizuoka
        </option>
        <option
          value="23"
        >
          Aichi
        </option>
        <option
          value="24"
        >
          Mie
        </option>
        <option
          value="25"
        >
          Shiga
        </option>
        <option
          value="26"
        >
          Kyoto
        </option>
        <option
          value="27"
        >
          Osaka
        </option>
        <option
          value="28"
        >
          Hyogo
        </option>
        <option
          value="29"
        >
          Nara
        </option>
        <option
          value="30"
        >
          Wakayama
        </option>
        <option
          value="31"
        >
          Tottori
        </option>
        <option
          value="32"
        >
          Shimane
        </option>
        <option
          value="33"
        >
          Okayama
        </option>
        <option
          value="34"
        >
          Hiroshima
        </option>
        <option
          value="35"
        >
          Yamaguchi
        </option>
        <option
          value="36"
        >
          Tokushima
        </option>
        <option
          value="37"
        >
          Kagawa
        </option>
        <option
          value="38"
        >
          Ehime
        </option>
        <option
          value="39"
        >
          Kochi
        </option>
        <option
          value="40"
        >
          Fukuoka
        </option>
        <option
          value="41"
        >
          Saga
        </option>
        <option
          value="42"
        >
          Nagasaki
        </option>
        <option
          value="43"
        >
          Kumamoto
        </option>
        <option
          value="44"
        >
          Oita
        </option>
        <option
          value="45"
        >
          Miyazaki
        </option>
        <option
          value="46"
        >
          Kagoshima
        </option>
        <option
          value="47"
        >
          Okinawa
        </option>
        <option
          value="48"
        >
          Hong Kong
        </option>
        <option
          value="49"
        >
          Korea
        </option>
        <option
          value="50"
        >
          Taiwan
        </option>
        <option
          value="51"
        >
          USA
        </option>
        <option
          value="52"
        >
          Europe
        </option>
        <option
          value="53"
        >
          Oversea
        </option>
        <option
          value="106"
        >
          Canada
        </option>
        <option
          value="107"
        >
          Singapore
        </option>
        <option
          value="108"
        >
          Thailand
        </option>
        <option
          value="109"
        >
          Australia
        </option>
        <option
          value="110"
        >
          New Zealand
        </option>
        <option
          value="111"
        >
          United Kingdom
        </option>
        <option
          value="112"
        >
          Italy
        </option>
        <option
          value="113"
        >
          Spain
        </option>
        <option
          value="114"
        >
          Germany
        </option>
        <option
          value="115"
        >
          France
        </option>
        <option
          value="116"
        >
          Portugal
        </option>
        <option
          value="117"
        >
          Indonesia
        </option>
        <option
          value="118"
        >
          Philippines
        </option>
        
        
      </select>
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="field"
  >
    
    <label
      class="label"
    >
      
      DDR CODE(optional)
      
    </label>
    
    
    <div
      class="control"
    >
      <input
        autocomplete="off"
        class="input"
        maxlength="8"
        minlength="8"
        modelmodifiers="[object Object]"
        pattern="^[1-9]\\\\d{7}$"
        placeholder="10000000"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="field"
  >
    
    <label
      class="label"
    >
      
      Password for Import(optional) 
      <div
        class="b-tooltip is-dark is-right"
      >
        <transition-stub
          appear="false"
          css="true"
          name="fade"
          persisted="false"
        >
          <div
            class="tooltip-content is-right is-dark"
            style="display: none;"
          >
            <span
              class=""
            />
            
            keep it blank if you do not use external import.
            
          </div>
        </transition-stub>
        <div
          class="tooltip-trigger"
        >
          
          <span
            class="icon is-small"
          >
            <i
              class="mdi mdi-help-circle-outline"
            />
          </span>
          
        </div>
      </div>
      
    </label>
    
    
    <div
      class="control has-icons-right"
    >
      <input
        autocomplete="off"
        class="input"
        type="password"
      />
      <!--v-if-->
      <span
        class="icon is-clickable is-right"
      >
        <i
          class="mdi mdi-eye mdi-24px"
        />
      </span>
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="field"
  >
    
    <!--v-if-->
    
    <div
      class="control"
    >
      <div
        class="field is-grouped-multiline is-grouped"
      >
        
        <label
          class="switch is-rounded"
        >
          <input
            class=""
            false-value="false"
            true-value="true"
            type="checkbox"
            value="false"
          />
          <span
            class="check"
          >
            <span
              class=""
            />
          </span>
          <span
            class="control-label"
          >
            
            Public
            
          </span>
        </label>
        <div
          class="help"
        >
          <p>
            You will be shown in the user search and your page will be visible to anyone.
            <br />
             Registered scores will be public.
          </p>
          <p
            class="has-text-weight-bold has-text-danger"
          >
            <span
              class="icon is-small"
            >
              <i
                class="mdi mdi-alert"
              />
            </span>
             Note: Even if you turn off, the scores so far will not be private.
          </p>
        </div>
        
      </div>
    </div>
    <!--v-if-->
  </div>
  <div
    class="field"
  >
    
    <!--v-if-->
    
    
    <button
      class="button is-success"
      disabled=""
      type="button"
    >
      <span
        class="button-wrapper"
      >
        <!--v-if-->
        <span
          class=""
        >
          
          Save
          
        </span>
        <!--v-if-->
      </span>
    </button>
    
    <!--v-if-->
  </div>
</section>
`;

exports[`Page /profile > { locale: "ja" } snapshot test > renders correctly 1`] = `
<section
  class="section"
>
  <transition-stub
    appear="false"
    css="true"
    name="fade"
    persisted="false"
  >
    <article
      class="notification is-danger"
    >
      <!--v-if-->
      <!--v-if-->
      <div
        class="media"
      >
        <span
          aria-hidden=""
          class="icon is-large media-left"
        >
          <i
            class="mdi mdi-alert-circle mdi-48px"
          />
        </span>
        <div
          class="media-content"
        >
          
          ユーザー登録は、まだ完了していません。引き続き、以下の情報を入力してください。
          
        </div>
      </div>
    </article>
  </transition-stub>
  <h1
    class="title"
  >
    ユーザー設定
  </h1>
  <div
    class="field"
  >
    
    <label
      class="label"
    >
      
      ID 
      <div
        class="b-tooltip is-dark is-right"
      >
        <transition-stub
          appear="false"
          css="true"
          name="fade"
          persisted="false"
        >
          <div
            class="tooltip-content is-right is-dark"
            style="display: none;"
          >
            <span
              class=""
            />
            
            登録後の変更はできません。
            
          </div>
        </transition-stub>
        <div
          class="tooltip-trigger"
        >
          
          <span
            class="icon is-small"
          >
            <i
              class="mdi mdi-help-circle-outline"
            />
          </span>
          
        </div>
      </div>
      
    </label>
    
    
    <div
      class="control"
    >
      <input
        autocomplete="off"
        class="input"
        loading="false"
        placeholder="ID"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="field"
  >
    
    <label
      class="label"
    >
      
      表示名
      
    </label>
    
    
    <div
      class="control"
    >
      <input
        autocomplete="off"
        class="input"
        required=""
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="field"
  >
    
    <label
      class="label"
    >
      
      所属地域 
      <div
        class="b-tooltip is-dark is-right"
      >
        <transition-stub
          appear="false"
          css="true"
          name="fade"
          persisted="false"
        >
          <div
            class="tooltip-content is-right is-dark"
            style="display: none;"
          >
            <span
              class=""
            />
            
            登録後の変更はできません。
            
          </div>
        </transition-stub>
        <div
          class="tooltip-trigger"
        >
          
          <span
            class="icon is-small"
          >
            <i
              class="mdi mdi-help-circle-outline"
            />
          </span>
          
        </div>
      </div>
      
    </label>
    
    
    <div
      class="control select"
    >
      <select
        class=""
        modelmodifiers="[object Object]"
      >
        
        <!--v-if-->
        
        
        
        <option
          value="0"
        >
          未指定
        </option>
        <option
          value="1"
        >
          北海道
        </option>
        <option
          value="2"
        >
          青森県
        </option>
        <option
          value="3"
        >
          岩手県
        </option>
        <option
          value="4"
        >
          宮城県
        </option>
        <option
          value="5"
        >
          秋田県
        </option>
        <option
          value="6"
        >
          山形県
        </option>
        <option
          value="7"
        >
          福島県
        </option>
        <option
          value="8"
        >
          茨城県
        </option>
        <option
          value="9"
        >
          栃木県
        </option>
        <option
          value="10"
        >
          群馬県
        </option>
        <option
          value="11"
        >
          埼玉県
        </option>
        <option
          value="12"
        >
          千葉県
        </option>
        <option
          value="13"
        >
          東京都
        </option>
        <option
          value="14"
        >
          神奈川県
        </option>
        <option
          value="15"
        >
          新潟県
        </option>
        <option
          value="16"
        >
          富山県
        </option>
        <option
          value="17"
        >
          石川県
        </option>
        <option
          value="18"
        >
          福井県
        </option>
        <option
          value="19"
        >
          山梨県
        </option>
        <option
          value="20"
        >
          長野県
        </option>
        <option
          value="21"
        >
          岐阜県
        </option>
        <option
          value="22"
        >
          静岡県
        </option>
        <option
          value="23"
        >
          愛知県
        </option>
        <option
          value="24"
        >
          三重県
        </option>
        <option
          value="25"
        >
          滋賀県
        </option>
        <option
          value="26"
        >
          京都府
        </option>
        <option
          value="27"
        >
          大阪府
        </option>
        <option
          value="28"
        >
          兵庫県
        </option>
        <option
          value="29"
        >
          奈良県
        </option>
        <option
          value="30"
        >
          和歌山県
        </option>
        <option
          value="31"
        >
          鳥取県
        </option>
        <option
          value="32"
        >
          島根県
        </option>
        <option
          value="33"
        >
          岡山県
        </option>
        <option
          value="34"
        >
          広島県
        </option>
        <option
          value="35"
        >
          山口県
        </option>
        <option
          value="36"
        >
          徳島県
        </option>
        <option
          value="37"
        >
          香川県
        </option>
        <option
          value="38"
        >
          愛媛県
        </option>
        <option
          value="39"
        >
          高知県
        </option>
        <option
          value="40"
        >
          福岡県
        </option>
        <option
          value="41"
        >
          佐賀県
        </option>
        <option
          value="42"
        >
          長崎県
        </option>
        <option
          value="43"
        >
          熊本県
        </option>
        <option
          value="44"
        >
          大分県
        </option>
        <option
          value="45"
        >
          宮崎県
        </option>
        <option
          value="46"
        >
          鹿児島県
        </option>
        <option
          value="47"
        >
          沖縄県
        </option>
        <option
          value="48"
        >
          香港
        </option>
        <option
          value="49"
        >
          韓国
        </option>
        <option
          value="50"
        >
          台湾
        </option>
        <option
          value="51"
        >
          アメリカ
        </option>
        <option
          value="52"
        >
          ヨーロッパ
        </option>
        <option
          value="53"
        >
          海外
        </option>
        <option
          value="106"
        >
          カナダ
        </option>
        <option
          value="107"
        >
          シンガポール
        </option>
        <option
          value="108"
        >
          タイ
        </option>
        <option
          value="109"
        >
          オーストラリア
        </option>
        <option
          value="110"
        >
          ニュージーランド
        </option>
        <option
          value="111"
        >
          イギリス
        </option>
        <option
          value="112"
        >
          イタリア
        </option>
        <option
          value="113"
        >
          スペイン
        </option>
        <option
          value="114"
        >
          ドイツ
        </option>
        <option
          value="115"
        >
          フランス
        </option>
        <option
          value="116"
        >
          ポルトガル
        </option>
        <option
          value="117"
        >
          インドネシア
        </option>
        <option
          value="118"
        >
          フィリピン
        </option>
        
        
      </select>
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="field"
  >
    
    <label
      class="label"
    >
      
      DDR CODE(任意)
      
    </label>
    
    
    <div
      class="control"
    >
      <input
        autocomplete="off"
        class="input"
        maxlength="8"
        minlength="8"
        modelmodifiers="[object Object]"
        pattern="^[1-9]\\\\d{7}$"
        placeholder="10000000"
        type="text"
      />
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="field"
  >
    
    <label
      class="label"
    >
      
      インポート用パスワード(任意) 
      <div
        class="b-tooltip is-dark is-right"
      >
        <transition-stub
          appear="false"
          css="true"
          name="fade"
          persisted="false"
        >
          <div
            class="tooltip-content is-right is-dark"
            style="display: none;"
          >
            <span
              class=""
            />
            
            外部インポートを利用しない場合は空欄にしてください。
            
          </div>
        </transition-stub>
        <div
          class="tooltip-trigger"
        >
          
          <span
            class="icon is-small"
          >
            <i
              class="mdi mdi-help-circle-outline"
            />
          </span>
          
        </div>
      </div>
      
    </label>
    
    
    <div
      class="control has-icons-right"
    >
      <input
        autocomplete="off"
        class="input"
        type="password"
      />
      <!--v-if-->
      <span
        class="icon is-clickable is-right"
      >
        <i
          class="mdi mdi-eye mdi-24px"
        />
      </span>
      <!--v-if-->
    </div>
    
    <!--v-if-->
  </div>
  <div
    class="field"
  >
    
    <!--v-if-->
    
    <div
      class="control"
    >
      <div
        class="field is-grouped-multiline is-grouped"
      >
        
        <label
          class="switch is-rounded"
        >
          <input
            class=""
            false-value="false"
            true-value="true"
            type="checkbox"
            value="false"
          />
          <span
            class="check"
          >
            <span
              class=""
            />
          </span>
          <span
            class="control-label"
          >
            
            公開する
            
          </span>
        </label>
        <div
          class="help"
        >
          <p>
            ユーザー検索に表示され、ユーザーページは誰でも閲覧できるようになります。
            <br />
             これから登録するスコアはランキングに自動登録され、一般公開されます。
          </p>
          <p
            class="has-text-weight-bold has-text-danger"
          >
            <span
              class="icon is-small"
            >
              <i
                class="mdi mdi-alert"
              />
            </span>
             ONからOFFにしても、今までに登録したスコアは非公開になりません。ご注意ください。
          </p>
        </div>
        
      </div>
    </div>
    <!--v-if-->
  </div>
  <div
    class="field"
  >
    
    <!--v-if-->
    
    
    <button
      class="button is-success"
      disabled=""
      type="button"
    >
      <span
        class="button-wrapper"
      >
        <!--v-if-->
        <span
          class=""
        >
          
          保存
          
        </span>
        <!--v-if-->
      </span>
    </button>
    
    <!--v-if-->
  </div>
</section>
`;
